File: macos/Sources/TCShellCore/RoutingCopy.swift
import Foundation

/// The routing surface's fixed words, decoded from what the Rust exported.
///
/// Every property here is filled from the payload. None of them has a
/// default and none is written in Swift: a word this shell invented would be
/// a word the Linux and Windows shells do not print, and `word_private` is a
/// privacy claim, so inventing one is inventing a claim.
///
/// Decoding is here rather than in `TCBridge` so it can be tested without
/// linking the dylib; `TCBridgeTests` checks the same properties against the
/// real export.
public struct RoutingCopy: Decodable, Equatable, Sendable {
    public let toolsHeading: String
    /// The one word on this surface that claims privacy.
    public let wordPrivate: String
    /// The not-wired word. Deliberately not "Not private": "Private" is a
    /// substring of that, and a shell matching with `contains` would find
    /// the wrong one.
    public let wordDirect: String
    public let wordUnknown: String
    public let wordNotUsed: String
    public let toolClaude: String
    public let toolCodex: String
    public let toolGemini: String
    public let intro: String
    public let toggle: String
    public let appliesAtOnce: String
    public let portTitle: String
    public let portNote: String
    public let folderTitle: String
    public let folderNote: String
    public let apply: String
    public let checking: String
    public let checkUnavailable: String
    public let probeReachable: String
    public let stateOff: String
    public let stateWaiting: String
    public let stateReading: String

    enum CodingKeys: String, CodingKey {
        case toolsHeading = "tools_heading"
        case wordPrivate = "word_private"
        case wordDirect = "word_direct"
        case wordUnknown = "word_unknown"
        case wordNotUsed = "word_not_used"
        case toolClaude = "tool_claude"
        case toolCodex = "tool_codex"
        case toolGemini = "tool_gemini"
        case intro
        case toggle
        case appliesAtOnce = "applies_at_once"
        case portTitle = "port_title"
        case portNote = "port_note"
        case folderTitle = "folder_title"
        case folderNote = "folder_note"
        case apply
        case checking
        case checkUnavailable = "check_unavailable"
        case probeReachable = "probe_reachable"
        case stateOff = "state_off"
        case stateWaiting = "state_waiting"
        case stateReading = "state_reading"
    }

    /// Decode the payload, or nil if it will not parse.
    ///
    /// Nil, never a partly-filled value with placeholder words: a screen that
    /// renders "" beside a tool name is worse than one that renders nothing,
    /// and a screen that renders a Swift-authored word is worse than both.
    public static func decode(fromJSON json: String) -> RoutingCopy? {
        guard let data = json.data(using: .utf8) else { return nil }
        return try? JSONDecoder().decode(RoutingCopy.self, from: data)
    }

    /// The four words, in the order the surface uses them.
    public var words: [String] {
        [wordPrivate, wordDirect, wordUnknown, wordNotUsed]
    }
}

File: windows/src/TraceCommons.Interop/RoutingCopy.cs
using System;
using System.Text.Json.Serialization;

namespace TraceCommons.Interop;

/// <summary>
/// The routing surface's fixed words, read from the Rust rather than kept
/// here.
///
/// Every property is filled from the payload and none has a default: a word
/// this shell invented would be a word the Linux and macOS shells do not
/// print, and <see cref="WordPrivate"/> is a privacy claim, so inventing one
/// is inventing a claim.
///
/// Exactly one word claims privacy and none denies it. Do not derive a "not
/// private" label from any of the others: "Private" is a substring of "Not
/// private", and a surface carrying both is one <c>Contains</c> away from
/// showing the wrong verdict.
/// </summary>
public sealed record RoutingCopy
{
    [JsonPropertyName("tools_heading")] public string ToolsHeading { get; init; } = "";

    /// <summary>The one word on this surface that claims privacy.</summary>
    [JsonPropertyName("word_private")] public string WordPrivate { get; init; } = "";

    /// <summary>
    /// The not-wired word. Deliberately not "Not private" -- see the type
    /// summary.
    /// </summary>
    [JsonPropertyName("word_direct")] public string WordDirect { get; init; } = "";

    [JsonPropertyName("word_unknown")] public string WordUnknown { get; init; } = "";
    [JsonPropertyName("word_not_used")] public string WordNotUsed { get; init; } = "";
    [JsonPropertyName("tool_claude")] public string ToolClaude { get; init; } = "";
    [JsonPropertyName("tool_codex")] public string ToolCodex { get; init; } = "";
    [JsonPropertyName("tool_gemini")] public string ToolGemini { get; init; } = "";
    [JsonPropertyName("intro")] public string Intro { get; init; } = "";
    [JsonPropertyName("toggle")] public string Toggle { get; init; } = "";
    [JsonPropertyName("applies_at_once")] public string AppliesAtOnce { get; init; } = "";
    [JsonPropertyName("port_title")] public string PortTitle { get; init; } = "";
    [JsonPropertyName("port_note")] public string PortNote { get; init; } = "";
    [JsonPropertyName("folder_title")] public string FolderTitle { get; init; } = "";
    [JsonPropertyName("folder_note")] public string FolderNote { get; init; } = "";
    [JsonPropertyName("apply")] public string Apply { get; init; } = "";
    [JsonPropertyName("checking")] public string Checking { get; init; } = "";
    [JsonPropertyName("check_unavailable")] public string CheckUnavailable { get; init; } = "";
    [JsonPropertyName("probe_reachable")] public string ProbeReachable { get; init; } = "";
    [JsonPropertyName("state_off")] public string StateOff { get; init; } = "";
    [JsonPropertyName("state_waiting")] public string StateWaiting { get; init; } = "";
    [JsonPropertyName("state_reading")] public string StateReading { get; init; } = "";

    /// <summary>The four words, in the order the surface uses them.</summary>
    public string[] Words => new[] { WordPrivate, WordDirect, WordUnknown, WordNotUsed };
}
